### PR TITLE
🏗✨ Adopt `prettier` for code formatting 

### LINKED ISSUE #13966: Include validator code in lint and presubmit checks
The code in validator is currently exempted from `gulp lint` and `gulp presubmit`. It's probably worth removing these exemptions, so that the validator code can benefit from all the new checks we've recently put in place (to the extent it makes sense to do so).

See https://github.com/ampproject/amphtml/blob/master/build-system/config.js#L125-L158

### LINKED ISSUE #17086: Adopt `prettier` for code formatting
https://github.com/prettier/prettier

### LINKED ISSUE #18601: Fix all remaining lint warnings in `validator/`
There are still about 200 lint warnings in `ampproject/amphtml`, and a large number of these are in `validator/`. This is a tracking bug to eventually eliminate all of them (or selectively disable rules that don't make sense).

See https://travis-ci.org/ampproject/amphtml/jobs/437867216#L660 for a list of validator warnings (search for `ampproject/amphtml/validator`)

### PR DESCRIPTION
**Pull request by rsimha**
This PR adopts [Prettier](https://prettier.io/) as the de-facto code formatting standard for the `ampproject/amphtml` codebase.

With this, the recommended developer workflow is to install the `eslint` plugin for your code editor, and set it to auto-fix `.js` files on save. (VSCode example [here](https://github.com/microsoft/vscode-eslint).)

**Changes, in order:**
- Installed the latest version of `prettier` as an `eslint` plugin
- Replaced all formatting rules in `.eslintrc` with `prettier`
- Moved all `/*OK*/` annotations associated with function calls to prefix the function name
- Fixed several incorrect `@type` annotations in AMP code
- Exempted `validator/` from linting, since it uses a separate set of rules internal to Google
- Ran Prettier against the entire codebase using `gulp lint --fix`
- Fixed all `no-useless-concat` lint errors that arose as a result of prettier's fixes
- Switched some `eslint-disable-line` annotations to `eslint-disable-next-line`
- Added documentation with editor recommendations and help for how to handle errors

**Workflow changes once this PR is merged:**
- **Option 1:** Use an editor with Eslint plugin support, and enable auto-fix on save. ([more info](https://github.com/rsimha/amphtml/blob/2019-03-01-Prettier/contributing/getting-started-e2e.md#code-quality-and-style))
- **Option 2:** Manually run `gulp lint --local-changes --fix` before uploading a commit.

Fixes #17086
Closes #18601
Closes #13966

### CODE REVIEW

**Comment by jridgewell** [comment]
The most egregious to me: https://github.com/prettier/prettier/issues/5806

**Comment by vjeux** [comment]
This is super exciting! Thanks for documenting all the issues you're facing during the migration.

As a quick workaround of the html template issue, I believe that you could call .trim() on the string and that would fix the issues.

**Comment by vjeux** [comment]
I'm curious, what tooling is using the `/*OK*/` annotation?

**Comment by jridgewell** [comment]
`/*OK*/` is our own annotation system. Originally, we used a primitive regex for `.foo`, and reported warnings. So, putting any comment between the `.` and `foo` made the regex not match.

Now, we use eslint and check the `leadingComments` on the property. But we've stuck with `/*OK*/` comments.

**Comment by rsimha** [comment]
We have a new Prettier release with bug fixes at https://prettier.io/blog/2019/04/12/1.17.0.html. (Yay!) I synced to `master`, re-ran the auto-fixer, and pushed a new commit at 1d990a536b46fb14384200f9015d73c33689626c.

Current status:

1. Quoted object keys are now being preserved (https://github.com/prettier/prettier/issues/4327) ✔️   
2. Inline `/*OK*/` tags are still being moved across parentheses ([link](https://github.com/ampproject/amphtml/pull/21212/commits/1d990a536b46fb14384200f9015d73c33689626c#diff-ce14342f14ddea0aae107c8fcf1159d3R121), https://github.com/prettier/prettier/issues/5973) ✖️  
3.  Typecast parentheses are still being moved incorrectly ([link](https://github.com/ampproject/amphtml/pull/21212/commits/1d990a536b46fb14384200f9015d73c33689626c#diff-a42a1edb64052bca9392c9dbd726d877R88), https://github.com/prettier/prettier/issues/4799#issuecomment-470322897) ✖️ 
4. HTML template strings are still being reformatted ([link](https://github.com/ampproject/amphtml/pull/21212/commits/1d990a536b46fb14384200f9015d73c33689626c#diff-cdf28404e8122e81dfbc43f1852dece1R83), https://github.com/prettier/prettier/issues/5588#issuecomment-470418654) ✖️ 

There may be ways to work around 2. and 4., but I think 3. is still a dealbreaker.

/to @cramforce @jridgewell @alanorozco @aghassemi for comment
/cc @vjeux for visibility

**Comment by rsimha** [comment]
I ran some checks on the new code.
- Item 2 (inline `/*OK*/` tags) affects just 4 or 5 function calls, so I think we can work around it. See the [output](https://gist.github.com/rsimha/10e0ba79692f1a2716d65beaef577dd3) from `gulp presubmit`
- Item 3 (typecast parentheses) is a blocker. The [output](https://gist.github.com/rsimha/732acbc367c0231ef9c091b50dfdb266) from `gulp check-types` detects only a subset of incorrect typecasts.
- Item 4 can be worked around by using `.trim()`, but we'd need a mechanism to prevent cases that aren't using it.

**Comment by jridgewell** [comment]
4 need not be fixed for us to adopt prettier, our code will work just fine. My dislike for it is only because it is an actual code change, not a formatting change.

I think only 3 needs to be fixed before we can merge this.

**Comment by rsimha** [comment]
**Update:** We have a new [Prettier release 1.17.1](https://github.com/prettier/prettier/blob/master/CHANGELOG.md#1171). With this, none of the issues mentioned in https://github.com/ampproject/amphtml/pull/21212#issuecomment-486542534 are blockers any more.

This PR has been updated with the latest version, and is now ready for review.

File: 3p/3p.js [hunk 1]
```diff
@@ -82,9 +79,7 @@ export function run(id, win, data) {
  * @param {function()=} opt_cb
  */
 export function writeScript(win, url, opt_cb) {
-  /*eslint no-useless-concat: 0*/
-  win.document
-      .write('<' + 'script src="' + encodeURI(url) + '"><' + '/script>');
+  win.document.write('<script src="' + encodeURI(url) + '"></script>');
```
Comment (cramforce) [inline 3p/3p.js]: This would be problematic if we inlined the script, because it would terminate the outer `<script>` element. But it is probably OK.

Comment (rsimha) [reply 3p/3p.js]: Good catch. String concatenations aren't automatically done by Prettier; this was the result of a find-replace done by me.

I've reverted the two string concatenations in this PR that contain a `<script>` tag, and exempted them using `eslint-disable-next-line`.

**Comment by rsimha** [comment]
This PR is now ready for a full review. Some notes:

- The overwhelming majority of changes come from the commit titled `First pass of prettier across entire codebase` (a4f6d76c67ba58cb32bff3ba963335d2c4d17680), which contains the automatic fixes applied by `prettier`. This commit can be skimmed through, since we've eliminated breaking changes in the above comments.
- The rest of the commits contain manual changes made by me (see PR description), and should be reviewed as usual. 

Tagging a few reviewers to distribute the load. Please post a comment if you see any objectionable changes in the code you own: @cramforce @aghassemi @choumx @newmuis @lannka @Gregable @jpettitt @kristoferbaxter @rsimha 

Once this is approved for merge, I will rebase the entire PR on the latest `HEAD`, rerun `prettier`, and make sure Travis is green before merging. (I'll likely do this late on Friday evening.)

**Comment by Enriqe** [comment]
@newmuis sure!

**Comment by choumx** [comment]
As a sanity check, did we try diffing the compiled binaries before/after this PR? I'd expect it to be the same modulo any manual changes you made.

**Comment by kristoferbaxter** [comment]
From a cursory glance, everything looks good here.

**Comment by rsimha** [comment]
> As a sanity check, did we try diffing the compiled binaries before/after this PR? I'd expect it to be the same modulo any manual changes you made.

Great idea! I just did this, and the only difference in `v0.js` other than version timestamps was one instance of a concatenated string in an error message (this was expected).

**Comment by choumx** [comment]
Nice! Might be worth doing the same for other `dist/` binaries and one or two extensions.

**Comment by rsimha** [comment]
> Nice! Might be worth doing the same for other `dist/` binaries and one or two extensions.

I did better, and compared every single one of the 154 minified JS binaries we build today. 127 of them were identical character-for-character. I reviewed every change in the remaining 27 binaries, and only found concatenation and whitespace changes in strings (both expected).

Screenshot from 2019-05-15 16-28-58

**Comment by cramforce** [comment]
That small output diff makes super happy. Great analysis!

**Comment by rsimha** [comment]
At long last, this PR is ready to merge. I've rebuilt the runtime and re-tested it to make sure that https://github.com/ampproject/amphtml/pull/21212#issuecomment-492813618 still holds true. All Travis tests are green. Merging this now, before any new conflicts appear. :crossed_fingers:

**Comment by rsimha** [comment]
For posterity:

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">With a cool 138,754 line diff <a href="https://twitter.com/AMPhtml?ref_src=twsrc%5Etfw">@AMPhtml</a> migrated to <a href="https://twitter.com/PrettierCode?ref_src=twsrc%5Etfw">@prettiercode</a> <a href="https://t.co/nT1YCTN4rC">https://t.co/nT1YCTN4rC</a><br><br>Huge thanks to the Prettier team and especially <a href="https://twitter.com/Vjeux?ref_src=twsrc%5Etfw">@Vjeux</a> for helping us along the way, making changes and accepting our PRs.</p>&mdash; Malte Ubl (@cramforce) <a href="https://twitter.com/cramforce/status/1129081248880533504?ref_src=twsrc%5Etfw">May 16, 2019</a></blockquote>

**Pull request merged**